os/fs/smartfs/smartfs_utils.c

```diff
@@ -1716,139 +1962,270 @@ int smartfs_countdirentries(struct smartfs_mountpt_s *fs, struct smartfs_entry_s
 }
 
 /****************************************************************************
- * Name: smartfs_truncatefile
+ * Name: smartfs_shrinkfile
  *
- * Description: Truncates an existing file on the device so that it occupies
- *              zero bytes and can be completely re-written.
+ * Description:
+ *   Shrink the size of an existing file to the specified length
  *
- ****************************************************************************/
-
-int smartfs_truncatefile(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *entry, FAR struct smartfs_ofile_s *sf)
+ *****************************************************************************/
+int smartfs_shrinkfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
 {
-	int ret;
-	uint16_t nextsector;
-	uint16_t sector;
-	struct smartfs_chain_header_s *header;
+	int ret = OK;
 	struct smart_read_write_s readwrite;
+	struct smartfs_chain_header_s *chainheader;
+	uint16_t nextsector;
+	uint16_t data_len;
+	uint16_t offset;
+	uint16_t last_data;
 
-	/* Walk through the directory's sectors and count entries */
+	/* 'data_len' represents length of DATA in a sector, the chainheader is not counted in file data */
+	data_len = fs->fs_llformat.availbytes - sizeof(struct smartfs_chain_header_s);
+	/* Number of bytes to be retained in the last sector after shrinking file */
+	last_data = (length % data_len);
 
-	nextsector = entry->firstsector;
-	header = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	/* Seek till point 'length' of the file, file pointer lies at position of requested 'length' now */
+	smartfs_seek_internal(fs, sf, length, SEEK_SET);
+	offset = sizeof(struct smartfs_chain_header_s) + last_data;
+	/* Keep as many sectors as needed and replace extra bytes in the last needed sector with ERASEDSTATE */
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	memset(&sf->buffer[offset], CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes - offset);
+	chainheader = (struct smartfs_chain_header_s *)sf->buffer;
+	nextsector = SMARTFS_NEXTSECTOR(chainheader);
+	*(uint16_t *)chainheader->nextsector = SMARTFS_ERASEDSTATE_16BIT;
+	*(uint16_t *)chainheader->used = last_data;
```

Comment:
Below is correct, I think, because sf->bytewritten should include correct information.
This will corrupt used value, because sf->bytewritten has another value.
```
	*(uint16_t *)chainheader->nextsector = SMARTFS_ERASEDSTATE_16BIT;
	*(uint16_t *)chainheader->used = SMARTFS_ERASEDSTATE_16BIT
	sf->byteswritten = last_data;
```

Reply:
Ok,
Smartfs_sync_internal will take care of used bytes.